### PR TITLE
perf(semanticTokens): avoid parsing all returned data

### PR DESCRIPTION
**Pull request by kevinhwang91**
There're several performance bottlenecks for the semantic token in all projects/files on client side:

Just take Neovim as an example, Vim is similar.

The data returned by the server will be all parsed and asked for total extmarks from Neovim.
If the data is large:
1. coc.nvim will be lagging to handle other services asked from Neovim;
2. Neovim must fetch all extmarks which may be blocking a few hundred milliseconds;
3. Neovim must encode the data also may be blocking;

Solution From this PR:
1. Only parse the relative semantic tokens format to the absolute format.
   This phase is a very small cost for coc.nvim;
2. Centered on the cursor, only render 3 times window height range from the absolute format
   pending semantic tokens, Neovim can understand the line number. 1 window
   height range is not enough, users who move up and down can feel the
   redraw frequently. Scroll the half-page down (`<C-d>`) on the topline
   will span 1.5 times window height range, scroll the half-page up is similar.
   In other words, to achieve less redraw must render 3 times window height
   if users press `<C-d>` or `<C-u>`;
3. Only diff 3 times window height range between semantic token and extmarks;

### CODE REVIEW

**Comment by chemzqm** [comment]
How to reproduce the issue?

**Comment by kevinhwang91** [comment]
> How to reproduce the issue?

Do you mean the performance issue? Enable semantic token for `c` and open `https://github.com/neovim/neovim/blob/master/src/nvim/eval.c`, doing some editing and undo, move the cursor, you will feel the lagging for Neovim. Especially, doing some global change like `Gread` in fugitive is more easily perceptible.

**Comment by kevinhwang91** [comment]
@chemzqm @fannheyward @oblitum I'm sorry to bother you. Have you tried this PR? Editing text or scrolling screen is smooth.

**Comment by kevinhwang91** [comment]
> @kevinhwang91 I can't reproduce lag using clangd with `src/nvim/eval.c`, could you try add some logs? For example:
> 
> ```ts
>     let d = Date.now()
>     if (SemanticTokens.is(result)) {
>       tokens = result.data
>     } else if (previousResult && Array.isArray(result.edits)) {
>       tokens = previousResult.tokens
>       result.edits.forEach(e => {
>         tokens.splice(e.start, e.deleteCount ? e.deleteCount : 0, ...e.data)
>       })
>     } else {
>       return
>     }
>     this.previousResults = { resultId: result.resultId, tokens, version }
>     let res = this.getHighlightItems(doc, tokens, legend)
>     logger.debug('len:', tokens.length, res.length)
>     logger.debug('cost:', Date.now() - d)
> ```

The snippet you provided is a small cost.

https://github.com/neoclide/coc.nvim/blob/72aeffd7abb9dda236e973a404e99e36c7c29a63/src/handler/semanticTokensHighlights/buffer.ts#L174-L193 cost nodejs nearly 40-80ms

https://github.com/neoclide/coc.nvim/blob/72aeffd7abb9dda236e973a404e99e36c7c29a63/src/window.ts#L551 block Neovim 150-300ms (without https://github.com/neoclide/coc.nvim/pull/3621)

**Comment by chemzqm** [comment]
> cost nodejs nearly 40-80ms

Don't know why your computer so slow, it's about 4~5ms on my computer.

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 1]
```diff
@@ -255,54 +329,61 @@ export default class SemanticTokensBuffer implements SyncItem {
   public async doHighlight(forceFull = false): Promise<void> {
     if (!this.enabled) return
     this.cancel()
-    let visible = await this.nvim.call('pumvisible')
+    let { nvim } = this
+    let visible = await nvim.call('pumvisible')
     if (visible) return
-    let tokenSource = this.tokenSource = new CancellationTokenSource()
+    let r = await nvim.call('coc#window#render_range', [this.bufnr]) as [number, number]
```
Comment (chemzqm) [inline src/handler/semanticTokensHighlights/buffer.ts:335]: The range could be changed (on CursorMoved) during request, so you'd better not request the range here.

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 2]
```diff
@@ -255,54 +329,61 @@ export default class SemanticTokensBuffer implements SyncItem {
...
     // request cancelled or can't work
-    if (!items || tokenSource.token.isCancellationRequested || this.invalid) return
-    let diff = await window.diffHighlights(this.bufnr, NAMESPACE, items)
+    // shouldRangeHighlight need check this.tokenSource
+    if (!items || tokenSource.token.isCancellationRequested || this.invalid) {
+      tokenSource.cancel()
+      this.tokenSource = null
+      return
+    }
+    let diff = await window.diffHighlights(this.bufnr, NAMESPACE, items, startLine, endLine)
```
Comment (chemzqm) [inline src/handler/semanticTokensHighlights/buffer.ts:362]: You can simply use `coc#highlight#update_highlights` for range update instead diff and apply, which would be faster. Further more, it's better to perform diff apply when the document not have any highlights applied (`coc#highlight#get_highlights` would not cause delay without highlight items).

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 3]
```diff
@@ -240,12 +283,43 @@ export default class SemanticTokensBuffer implements SyncItem {
...
+    }
+  }
+
+  public async doRangeHighlight(startLine?: number, endLine?: number): Promise<void> {
     if (!this.enabled) return
+    if (!startLine || !endLine) {
+      let r = await this.nvim.call('coc#window#render_range', [this.bufnr]) as [number, number]
+      if (!r) return
+      [startLine, endLine] = r
+    }
```
Comment (chemzqm) [inline src/handler/semanticTokensHighlights/buffer.ts:318]: Not looks good since it's not guard by CancellationTokenSource any more.

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 4]
```diff
@@ -240,12 +283,43 @@ export default class SemanticTokensBuffer implements SyncItem {
...
+    let r = await this.nvim.call('coc#window#render_range', [this.bufnr]) as [number, number]
+    if (!r) return
+    let [startLine, endLine] = r
+    const items: HighlightItem[] = []
+    let doc = workspace.getDocument(this.bufnr)
+    let start: number | undefined
+    let end: number | undefined
+    for (let i = startLine - 1; i < endLine; i++) {
+      let arr = this._pendingHighlights.get(i)
+      if (!this._pendingHighlights.delete(i) || arr.length == 0) continue
```
Comment (chemzqm) [inline src/handler/semanticTokensHighlights/buffer.ts:297]: This delete could make further `coc#highlight#update_highlights` bug since it requires all highlight items in that range, but you may lost them by delete on previous `doPendingHighlight `

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 5]
```diff
@@ -124,7 +140,21 @@ export default class SemanticTokensBuffer implements SyncItem {
    * Get current highlight items
    */
   public get highlights(): ReadonlyArray<SemanticTokenRange> {
-    return this._highlights
+    let pending: SemanticTokenRange[] = []
+    this._pendingHighlights.forEach(absTokens => {
```
Comment (chemzqm) [inline src/handler/semanticTokensHighlights/buffer.ts:144]: `semanticTokens.inspect` requires all current highlights, but `_pendingHighlights` changes from time to time.

**Review by chemzqm — COMMENTED** [review]
你先别弄了，我这里重新做些优化吧

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 6]
```diff
@@ -353,7 +353,12 @@ export default class SemanticTokensBuffer implements SyncItem {
       items = await this.requestAllHighlights(startLine, endLine, tokenSource.token, forceFull)
     }
     // request cancelled or can't work
-    if (!items || tokenSource.token.isCancellationRequested || this.invalid) return
+    // shouldRangeHighlight need check this.tokenSource
+    if (!items || tokenSource.token.isCancellationRequested || this.invalid) {
+      tokenSource.cancel()
+      this.tokenSource = null
```
Comment (chemzqm) [inline src/handler/semanticTokensHighlights/buffer.ts:359]: You can't simply reset `this.tokenSource` since it could be changed during request

File: src/handler/semanticTokensHighlights/buffer.ts [hunk 7]
```diff
@@ -111,10 +123,14 @@ export default class SemanticTokensBuffer implements SyncItem {
 
   public get shouldRangeHighlight(): boolean {
     let doc = workspace.getDocument(this.bufnr)
-    if (!doc) return false
+    if (!doc || this.tokenSource) return false
```
[the remark this answers is not in the document]
Comment (kevinhwang91) [reply src/handler/semanticTokensHighlights/buffer.ts:126]: I don't think 

> 你先别弄了，我这里重新做些优化吧

ok

**Comment by kevinhwang91** [comment]
> 你先别弄了，我这里重新做些优化吧

好

File: src/handler/semanticTokensHighlights/buffer.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (chemzqm) [reply src/handler/semanticTokensHighlights/buffer.ts:126]: `requestAllHighlights` could be canceled or return `null`, range highlight should not be canceled on that case.
